### DISCUSSION TITLE
proposal: add support for ent/schema versioning

### DISCUSSION THREAD

**Discussion by a8m** [ent/ent#1487] [Ideas]
In this proposal, we suggest adding new functionality to the code-generation phase to compare the "current version" of the ent/schema with its "previous version".

This new capability will help the code-generator to give feedback on the new schema changes. For example, when a new field is added (by the user) to an existing schema, ent will suggest you add it with a default value, or make it optional if the table is not empty because this will cause the migration to fail. 

It will also make the migration "smarter", as there's no need to inspect the database since ent will know exactly what are the new changes it needs to execute - An additional proposal will be posted with the new migration API. 

### Implementation

The API will support 2 options for getting the previous schema version:

1. Branch name, commit, or a tag: `master`, `main`, `e3a586bd` or `v0.1.0`. 
2. File path or a URL: `./migrate/versions/..` or `https://github.com/ent/.../versions/...`

After adding the new migration API, the code-generator will be able to read it automatically from the `migrate/versions` directory.

The previous schema version is also going to be added to the `entc/gen` types, in order to expose this functionality to external tools like `entproto`. The API will provide helpers for getting the previous version of a field, edge, type, etc.

Please share with us your feedback, suggestions, or just random thoughts to help improve users' experience. I hope to share more details on the implementation after start working on it.

**Comment by rotemtam** [comment]
> optional if the table is not empty because this will cause the migration to fail.


This requires some more thought, consider that the same application can run multiple times (different installations, environments, etc.) Not sure schema versioning should be aware of a concrete backing database. That information makes sense in the context of the migration planner which I think is a separate component, WDYT?

**Comment by michaelkub** [comment]
> Branch name, commit, or a tag: master, main, e3a586bd or v0.1.0.
> File path or a URL: ./migrate/versions/.. or https://github.com/ent/.../versions/...

Will ent keep history of upgrades in database, something like upgrade table with date, schema version, etc... ? Locks on this table can be used in order to prevent  parallel upgrades
Not sure is it good or not, but can be useful ,generate migration script for every upgrade, it can allow to edit it and add some manual changes.
And last and not least, I need some way to fill data between upgrades,  so it should be as part of the upgrade process

**Reply by a8m** [reply]
Hey @michaelkub,
All these points are going to be addressed in the next release. I'll be happy to discuss it and get feedback in a meeting if you have time - I'll ping you in our Slack channel.

**Comment by Yiling-J** [comment]
I like this proposal, this sounds like Django migrations. Also because we always run offline migration on our prod server, don't need to scan database is a huge improvement.

**Comment by wangzuo** [comment]
I am currently using rails-like migrator (up & down sql) with a schema version table. Will it be possible if ent can automatically generate the up & down sql when schema changes? I find that i can generate the up sql with current api but not for the down sql.

**Reply by rotemtam** [reply]
Hi @wangzuo,
Thanks for your reply. 

Can you elaborate more on the migrator you are currently using? Is it an open-source/commercial/homegrown tool? How are you generating the `up` SQL? 

Regarding the future of migrations with `ent` - the likely path will be:
* Replace the current "online" migration with a new (much more robust) implementation (next version of ent)
* Create an option to explicitly generate migration files (using a specialized DSL) that will contain both up and down directions. (Soon after)
* Augment the migration driver with intelligent suggestions etc as @a8m mentioned above.

**Reply by wangzuo** [reply]
Thank you for the helpful reply, right now i guess`client.Schema.WriteTo` can be used to generate the "up" sql?  The migrator I am currently using is based on [popup migrator](https://github.com/gobuffalo/pop/blob/master/migrator.go), which has both DSL and plain sql support (i use sql). My plan to add `ent` is that `ent` outputs schema changes in sql (up & down) and keep using the existing migrator (not all migrations are schema changes). But I like the path you mentioned and will follow the development.
